[DDP] Make compute_bucket_assignment_by_size return per bucket sizes

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62279
* #62232
* __->__ #62231
* #62230
* #62229
* #62209

`compute_bucket_assignment_by_size` is responsible for setting per-bucket size limits, return this information from the function so that we are aware of size limits for each bucket.

This is currently not being consumed, but will be in the next diff when we log bucket size limits to DDP logging. This will help us run experiments under different bucket size configs and analyze the impact.

Differential Revision: [D29919056](https://our.internmc.facebook.com/intern/diff/D29919056/)

rohan-varma committed:
[DDP] Make compute_bucket_assignment_by_size return per bucket sizes

`compute_bucket_assignment_by_size` is responsible for setting per-bucket size limits, return this information from the function so that we are aware of size limits for each bucket.

This is currently not being consumed, but will be in the next diff when we log bucket size limits to DDP logging. This will help us run experiments under different bucket size configs and analyze the impact.

Differential Revision: [D29919056](https://our.internmc.facebook.com/intern/diff/D29919056/)

[ghstack-poisoned]

mrshenli commented on test/distributed/test_c10d_common.py:
```diff
@@ -515,8 +518,9 @@ def test_multi_limit_multi_dtype(self):
-        result = dist._compute_bucket_assignment_by_size(tensors, [200, 400])
+        result, per_bucket_size_limits = dist._compute_bucket_assignment_by_size(tensors, [200, 400])
         self.assertEqual([[0], [1], [2, 4], [3, 5]], result)
+        self.assertEqual(per_bucket_size_limits, [200, 200, 400, 400])
```
(this is prior to this PR) curious, why do we need to apply size limits independently to each dtype? IIRC, the original intention is to have a small first bucket to kick off comm earlier. If so, would it be sufficient to have just one small bucket followed by all large bucket limits? Or is this just to simply the implementation?

rohan-varma replied:
> IIRC, the original intention is to have a small first bucket to kick off comm earlier

I think this is my understanding as well, but interestingly before rebuild buckets, the first bucket size actually ends up being the last bucket size because we reverse parameter ready order when passing into the reducer. So prior to rebuild buckets PR landing, we actually didn't have a smaller first bucket size according to my understanding.

@zhaojuanmao pointed out this inconsistency and potentially we should investigate it more to determine where the smaller bucket makes the most sense. 

As far as I understand, the reason we apply size limits independently to different types is so that tensors with different types go into different buckets? At least that's what I can tell from `compute_bucket_assignment_by_size`, but not sure what the motivation for that is.

mrshenli commented on torch/csrc/distributed/c10d/comm.cpp:
```diff
@@ -61,9 +61,12 @@ void broadcast_coalesced(
   // This routine is multi-device aware, so the tensors can be split across
   // multiple devices and can contain a mix of CPU and CUDA tensors.
-  const auto buckets =
+  const auto tup =
```
shall we make this var name more readable? (e.g., `buckets_and_size_limits`) Or we can directly run `std::tie` here?

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -1928,7 +1934,7 @@ std::vector<std::vector<size_t>> compute_bucket_assignment_by_size(
     if (!expect_sparse_gradient.empty() &&
         expect_sparse_gradient[tensor_index]) {
-      result.push_back({tensor_index});
+          result.emplace_back(std::vector<size_t>({tensor_index}), kNoSizeLimit);
```
wrong indent?

rohan-varma replied:
Actually looks like the indent was off previously and on next line, at least according to linter

rohan-varma committed:
Update on "[DDP] Make compute_bucket_assignment_by_size return per bucket sizes"


`compute_bucket_assignment_by_size` is responsible for setting per-bucket size limits, return this information from the function so that we are aware of size limits for each bucket.

This is currently not being consumed, but will be in the next diff when we log bucket size limits to DDP logging. This will help us run experiments under different bucket size configs and analyze the impact.

Differential Revision: [D29919056](https://our.internmc.facebook.com/intern/diff/D29919056/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Make compute_bucket_assignment_by_size return per bucket sizes"


`compute_bucket_assignment_by_size` is responsible for setting per-bucket size limits, return this information from the function so that we are aware of size limits for each bucket.

This is currently not being consumed, but will be in the next diff when we log bucket size limits to DDP logging. This will help us run experiments under different bucket size configs and analyze the impact.

Differential Revision: [D29919056](https://our.internmc.facebook.com/intern/diff/D29919056/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Make compute_bucket_assignment_by_size return per bucket sizes"


`compute_bucket_assignment_by_size` is responsible for setting per-bucket size limits, return this information from the function so that we are aware of size limits for each bucket.

This is currently not being consumed, but will be in the next diff when we log bucket size limits to DDP logging. This will help us run experiments under different bucket size configs and analyze the impact.

Differential Revision: [D29919056](https://our.internmc.facebook.com/intern/diff/D29919056/)

[ghstack-poisoned]
